lib/rack/multipart/parser.rb

```diff
@@ -27,16 +27,19 @@ class BoundaryTooLongError < StandardError
       include BadRequest
     end
 
+    # Line folding not allowed in multipart headers
+    class ObsoleteFoldingError < StandardError
+      include BadRequest
+    end
+
     # Prefer to use the BoundaryTooLongError class or Rack::BadRequest.
     Error = BoundaryTooLongError
 
     EOL = "\r\n"
-    FWS = /[ \t]+(?:\r\n[ \t]+)?/ # whitespace with optional folding
-    HEADER_VALUE = "(?:[^\r\n]|\r\n[ \t])*" # anything but a non-folding CRLF
     MULTIPART = %r|\Amultipart/.*boundary=\"?([^\";,]+)\"?|ni
-    MULTIPART_CONTENT_TYPE = /^Content-Type:#{FWS}?(#{HEADER_VALUE})/ni
-    MULTIPART_CONTENT_DISPOSITION = /^Content-Disposition:#{FWS}?(#{HEADER_VALUE})/ni
-    MULTIPART_CONTENT_ID = /^Content-ID:#{FWS}?(#{HEADER_VALUE})/ni
+    MULTIPART_CONTENT_TYPE = /^Content-Type:[ \t]*(.*)#{EOL}/ni
+    MULTIPART_CONTENT_DISPOSITION = /^Content-Disposition:(.*)(?=#{EOL}(\S|\z))/ni
+    MULTIPART_CONTENT_ID = /^Content-ID:[ \t]*([^#{EOL}]*)/ni
```

Comment:
I don't have an opinion on treatment of line-folding, but this introduces an obvious ReDoS and we must not merge it as-is!